config.go

```diff
@@ -40,287 +38,123 @@ const (
 
 	// ServiceOwnerKey is the config key for a service owner.
 	ServiceOwnerKey = "owner"
-)
 
-const (
-	_appRoot     = "_ROOT"
-	_environment = "_ENVIRONMENT"
-	_configDir   = "_CONFIG_DIR"
-	_baseFile    = "base.yaml"
-	_secretsFile = "secrets.yaml"
-	_devEnv      = "development"
+	// _defaultInit is used to load a default set of configuration files.
+	_defaultInit = `
+- path: ${CONFIG_DIR:config}/base.yaml
+  interpolate: true
+- path: ${CONFIG_DIR:config}/${ENVIRONMENT:development}.yaml
+  interpolate: true
+- path: ${CONFIG_DIR:config}/secrets.yaml
+  optional: true
+`
+	// _testInit is used to specify files for the TestLoader.
+	_testInit = `
+- path: ${CONFIG_DIR:config}/base.yaml
+  interpolate: true
+- path: ${CONFIG_DIR:config}/test.yaml
+  interpolate: true
+`
 )
 
-type lookUpFunc func(string) (string, bool)
+// LookupFunc is a type alias for a function to look for environment variables,
+type LookUpFunc func(string) (string, bool)
+
+// ProviderComposition is a collection of functions to fold the configs.
+type ProviderComposition []func(LookUpFunc, Provider) (Provider, error)
 
 // Loader is responsible for loading config providers.
 type Loader struct {
-	lock sync.RWMutex
-
-	envPrefix            string
-	staticProviderFuncs  []ProviderFunc
-	dynamicProviderFuncs []DynamicProviderFunc
-
-	// Files to load, they will be replaced with values from environment variables.
-	configFiles []string
+	// LookUp is a function to look for interpolated/environment values
+	LookUp LookUpFunc
 
-	// Files to load without interpolation.
-	staticFiles []string
+	// Initial config to apply fold functions
+	Init Provider
 
-	// Dirs to load from.
-	dirs []string
-
-	// Where to look for environment variables.
-	lookUp lookUpFunc
+	// Functions to apply on initial provider.
+	Apply ProviderComposition
 }
 
 // DefaultLoader is going to be used by a service if config is not specified.
 // First values are going to be looked in dynamic providers, then in command line provider
 // and YAML provider is going to be the last.
-var DefaultLoader = NewLoader(commandLineProviderFunc)
-
-// NewLoader returns a default Loader with providers overriding the YAML provider.
-func NewLoader(providers ...ProviderFunc) *Loader {
-	l := &Loader{
-		envPrefix: "APP",
-		dirs:      []string{".", "./config"},
-		lookUp:    os.LookupEnv,
-	}
-
-	// Order is important: we want users to be able to override static provider
-	l.RegisterProviders(l.YamlProvider())
-	l.RegisterProviders(providers...)
-
-	return l
+var DefaultLoader = Loader{
+	LookUp: os.LookupEnv,
+	Init:   NewYAMLProviderFromReaderWithExpand(os.LookupEnv, ioutil.NopCloser(strings.NewReader(_defaultInit))),
```

Comment:
This seems bizarre. Why are we using yaml to express loading yaml?
Why can't this be `[]struct{path string; interpolate, optional bool}`?